### ISSUE TITLE
✨ Proposal - Add `setReturnFocus` option to `deactiave` method

### ISSUE THREAD

**Issue by far-fetched** [focus-trap/focus-trap#485]
### What and why:
In the time of creation `focus-trap` sometimes we don't know beforehand what element should be focused after deactivation. Imagine a situation with modal and focus-trap inside. When I click outside of modal (on focusable element) I would like to deactivate trap (modal is closed) and focus on clicked element.

### How:
We don't have access to config after creation of trap, but we have `deactiaveOptions` which are passed to `deactivate` method. Inside it, we can extend current behavior. In `finishDeactivation` method, focus element should be (order makes difference):
1. from `setReturnFocus` property passed to `deactivate` method
2. from `setReturnFocus` property passed to config
3. last focused node before activation

Willing to see any alternatives.
If the above plan makes sense, I can try to bake PR.

**Comment by stefcameron** [comment]
@far-fetched Thank you for this proposal, and for taking the time to look at the code to see what could be done. This is similar to #481.

I'm wondering, though, if this might already be possible. While the `setReturnFocus` option is set on trap creation (when you may not know what to return focus to), if you provide a __function__ to it, it won't be called until the trap is deactivated -- in the `finishDeactivation` function you point out:

1. If the trap is configured to return focus after deactivation, then [line 567](https://github.com/focus-trap/focus-trap/blob/master/index.js#L567) is executed.
2. Then [line 235](https://github.com/focus-trap/focus-trap/blob/master/index.js#L235) -- which is where the `setReturnFocus` option is checked, and if it's a function, called to get the node to return focus to.

That order of execution is definitely not clear in the README. I'll clarify that.

Do you still see a need for your feature request?

**Comment by stefcameron** [comment]
See #486

**Comment by far-fetched** [comment]
Hey, during implementation I struggle with understanding what is desired behavior...
You said that recently, `initial-focus()` was extended to accept false as return value when we don't want any initial focus. 
In code, there is only option to pass boolean value false:
[code](https://github.com/focus-trap/focus-trap/blob/master/index.js#L160)
I check how it behaves with `() => return false` function, and it seems that it throws error.
I have a hunch that the relevant line should be:
```js
let initialFocus = getOption({}, 'initialFocus');
if (initialFocs === false || (typeof initialFocs === 'function' && initialFocus() === false)
```
I understand that you wanted to implement similar behavior for `setReturnFocus`
Can you shed some light on this code ? Is current implementation correct ?

**Comment by stefcameron** [comment]
@far-fetched You're correct: The current implementation of the `initialFocus` option is incorrect. It should accept a function that returns false as well as just being false. The docs are incorrect as well because the docs indicate only the function can return false.

The fix, though, would need to be a bit different from what you suggest because what you're suggesting would result in the option-as-a-function being called twice because of line 164, so we should avoid that.

I'll fix this as a separate issue since it's not related to this issue which you're trying to fix.

**Comment by stefcameron** [comment]
@far-fetched I opened #491 to fix the `initialFocus` issue. I think you'll want to base your PR for `setReturnFocus` onto that branch, or wait for me to merge it to master and rebase. If you have a moment, please check it out and LMK if you think it looks good.

**Comment by far-fetched** [comment]
Thanks, looks good to me 🍰. I only left one minor comment about the doc.

**Comment by stefcameron** [comment]
@far-fetched 
> Thanks, looks good to me 🍰. I only left one minor comment about the doc.

Thanks for the comment. I saw it, you said, "If null, the function will throw error - even in code there is comment (line 153). IMHO null shouldn't be here in doc." -- but I can't find it on my PR for some reason. When I try to view your comment, GitHub tells me it was made on a commit that isn't part of my PR. So I can't find what line you're referring to. I thought it might have been [line 137 in index.js in my PR](https://github.com/focus-trap/focus-trap/pull/491/files#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346R137), but I don't see a reference to `null` as a return value there. Can you point me to the right place?

**Comment by far-fetched** [comment]
Hey I was referring to this [line](https://github.com/focus-trap/focus-trap/pull/491/files#diff-a3ce29c1c993dbf3f968461bb3ff5e3f522d8b0c94cc2ca8f6b3ef7a9eda3621R193)

**Comment by stefcameron** [comment]
> Hey I was referring to this [line](https://github.com/focus-trap/focus-trap/pull/491/files#diff-a3ce29c1c993dbf3f968461bb3ff5e3f522d8b0c94cc2ca8f6b3ef7a9eda3621R193)

Ah, OK, in the generated demo bundle. The actual code (in `index.js`) had been updated, but I hadn't re-build the library _and then_ re-generated the bundle since removing `null` from the return types. Glad you caught that. I updated the `yarn demo-bundle` package script to always run the CJS build prior to generating the bundle so that it does so with the latest `index.js` code: https://github.com/focus-trap/focus-trap/commit/ccc11ad3ce8e19dfb5c5d088bd006175ffbbe039

I have merged my PR.
